brave/src/main/java/brave/propagation/ExtraFields.java

```diff
@@ -0,0 +1,180 @@
+    return elements != null ? elements[index] : null;
+  }
+
+  @Override protected void forEach(FieldConsumer<ExtraField, String> fieldConsumer) {
+    String[] elements = values;
+    if (elements == null) return;
+
+    for (int i = 0, length = fields.length; i < length; i++) {
+      String value = elements[i];
+      if (value == null) continue;
+      fieldConsumer.accept(fields[i], value);
+    }
+  }
+
+  @Override protected final void put(ExtraField field, String value) {
+    int index = indexOf(field);
+    if (index == -1) return;
+    put(index, value);
+  }
+
+  @Override protected boolean isEmpty() {
+    String[] elements = values;
+    if (elements == null) return true;
+    for (String value : elements) {
+      if (value != null) return false;
+    }
+    return true;
+  }
+
+  protected final void put(int index, String value) {
+    if (index >= fields.length) return;
+
+    synchronized (this) {
+      doPut(index, value);
+    }
+  }
+
+  void doPut(int index, String value) {
+    String[] elements = values;
+    if (elements == null) {
+      elements = new String[fields.length];
+      elements[index] = value;
+    } else if (value.equals(elements[index])) {
+      return;
+    } else { // this is the copy-on-write part
+      elements = Arrays.copyOf(elements, elements.length);
+      elements[index] = value;
+    }
+    values = elements;
+  }
+
+  @Override protected final void putAllIfAbsent(PropagationFields parent) {
+    if (!(parent instanceof ExtraFields)) return;
+    ExtraFields predefinedParent = (ExtraFields) parent;
+    checkSameFields(predefinedParent);
+    String[] parentValues = predefinedParent.values;
+    if (parentValues == null) return;
+    for (int i = 0; i < parentValues.length; i++) {
+      if (parentValues[i] != null && get(i) == null) { // extracted wins vs parent
+        doPut(i, parentValues[i]);
+      }
+    }
+  }
+
+  void checkSameFields(ExtraFields predefinedParent) {
+    if (!Arrays.equals(fields, predefinedParent.fields)) {
+      throw new IllegalStateException(
+        String.format("Mixed name configuration unsupported: found %s, expected %s",
+          Arrays.toString(fields), Arrays.toString(predefinedParent.fields))
+      );
+    }
+  }
+
+  @Override public final Map<String, String> toMap() {
+    String[] elements = values;
+    if (elements == null) return Collections.emptyMap();
+
+    MapFieldConsumer result = new MapFieldConsumer();
+    forEach(result);
+    return result;
+  }
+
+  static final class MapFieldConsumer extends LinkedHashMap<String, String>
+    implements FieldConsumer<ExtraField, String> {
+    @Override public void accept(ExtraField field, String value) {
+      put(field.name, value);
+    }
+  }
+
+  int indexOf(ExtraField name) {
+    for (int i = 0, length = fields.length; i < length; i++) {
+      if (fields[i].equals(name)) return i;
+    }
+    return -1;
+  }
+
+  @Override public int hashCode() { // for unit tests
```

Comment:
Even only for unit tests, probably good to compare `fields` too or if a newly added test doesn't work because of that in the future it will be quite mysterious

Reply:
went offline for a while as I really don't want to do this.. I'm rejigging the tests instead.

Reply:
glad I didn't want to change the hashCode/equals.. there's a bug in the impl :)